org.eclipse.scanning.points/src/org/eclipse/scanning/points/CompoundSpgIteratorFactory.java

```diff
@@ -0,0 +1,173 @@
+			String[] regionAxes, PyObject[] mutators, int duration, boolean continuous) {
+		// TODO Matt D. 2017-10-25: where does this method belong?
+		final JythonObjectFactory<PyObject> excluderFactory = ScanPointGeneratorFactory.JExcluderFactory();
+		final JythonObjectFactory<ScanPointIterator> cpgFactory = ScanPointGeneratorFactory.JCompoundGeneratorFactory();
+		final List<PyObject> pyRegions = Arrays.asList(regions)
+				.stream()
+				.map(AbstractScanPointIterator::makePyRoi)
+				.filter(Objects::nonNull)
+				.collect(Collectors.toList());
+		final PyObject excluder = excluderFactory.createObject(pyRegions.toArray(), new PyList(Arrays.asList(regionAxes)));
+		final PyObject[] excluders = pyRegions.isEmpty() ? EMPTY_PY_ARRAY : new PyObject[] { excluder };
+		return cpgFactory.createObject(iterators, excluders, mutators, duration, continuous);
+	}
+
+	public CompoundSgpIterator createCompoundSpgIterator(CompoundGenerator gen) {
+		Iterator<? extends IPosition>[] iterators = initIterators(gen);
+
+		if (Arrays.stream(iterators).map(Object::getClass).anyMatch(IDeviceDependentIterable.class::isAssignableFrom)) {
+			throw new IllegalArgumentException();
+		}
+
+		final JythonObjectFactory<ScanPointIterator> compoundGeneratorFactory = ScanPointGeneratorFactory.JCompoundGeneratorFactory();
+
+	    final Object[] excluders = getExcluders(gen.getModel().getRegions());
+	    final Object[] mutators = getMutators(gen.getModel().getMutators());
+	    final double duration = gen.getModel().getDuration();
+	    final boolean continuous = isContinuous(gen.getModel());
+
+	    final ScanPointIterator pyIterator = compoundGeneratorFactory.createObject(
+				iterators, excluders, mutators, duration, continuous);
+
+	    return new CompoundSgpIterator(pyIterator);
+	}
+
+	/**
+	 * A Compound model is continuous if its inner most model is a map model and that model is continuous.
+	 * @return
+	 */
+	private static boolean isContinuous(CompoundModel<?> model) {
+		List<Object> models = model.getModels();
+		if (!models.isEmpty()) {
+			final Object innerModel = models.get(models.size() - 1);
+			return innerModel instanceof IMapPathModel && ((IMapPathModel) innerModel).isContinuous();
+		}
+		return false;
+	}
+
+	@SuppressWarnings("unchecked")
+	private static Iterator<? extends IPosition>[] initIterators(CompoundGenerator gen) {
+		return Arrays.stream(gen.getGenerators()).
+				map(IPointGenerator::iterator).
+				toArray(Iterator[]::new);
+	}
+
+	/**
+	 * Creates an array of python objects representing the mutators
+	 * @param mutators
+	 * @return
+	 */
+	private static Object[] getMutators(Collection<IMutator> mutators) {
+		if (mutators != null) {
+			return mutators.stream().map(IMutator::getMutatorAsJythonObject).toArray();
+		}
+
+		return new Object[0];
+	}
+
+	/**
+	 * Creates an array of python objects representing the excluders
+	 * @param regions
+	 * @return
+	 */
+	public static Object[] getExcluders(Collection<?> regions) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 19 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)